compiler/z/codegen/S390Peephole.cpp

```diff
@@ -0,0 +1,3248 @@
+#include "ras/Delimiter.hpp"                        // for Delimiter
+#include "runtime/Runtime.hpp"                      // for TR_LinkageInfo
+#include "z/codegen/CallSnippet.hpp"
+#include "z/codegen/OpMemToMem.hpp"
+#include "z/codegen/S390Evaluator.hpp"
+#include "z/codegen/S390GenerateInstructions.hpp"
+#include "z/codegen/S390Instruction.hpp"
+#include "z/codegen/S390OutOfLineCodeSection.hpp"
+#include "z/codegen/TRSystemLinkage.hpp"
+
+TR_S390Peephole::TR_S390Peephole(TR::Compilation* comp, TR::CodeGenerator *cg)
+   : _fe(comp->fe()),
+     _outFile(comp->getOutFile()),
+     _cursor(cg->getFirstInstruction()),
+     _cg(cg)
+   {
+   }
+
+TR_S390PreRAPeephole::TR_S390PreRAPeephole(TR::Compilation* comp, TR::CodeGenerator *cg)
+   : TR_S390Peephole(comp, cg)
+   {
+   }
+
+void
+TR_S390PreRAPeephole::perform()
+   {
+   TR::Delimiter d(comp(), comp()->getOption(TR_TraceCG), "preRAPeephole");
+
+   if (comp()->getOption(TR_TraceCG))
+      printInfo("\nPost PreRegister Assignment Peephole Optimization Instructions:\n");
+
+   bool moveInstr;
+
+   while (_cursor != NULL)
+      {
+      switch(_cursor->getOpCodeValue())
+         {
+         case TR::InstOpCode::L:
+            {
+            LoadStoreReduction(TR::InstOpCode::ST, 4);
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         case TR::InstOpCode::LFH:
+            {
+            LoadStoreReduction(TR::InstOpCode::STFH, 4);
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         case TR::InstOpCode::LG:
+            {
+            LoadStoreReduction(TR::InstOpCode::STG, 8);
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         default:
+            {
+            if (comp()->getOption(TR_TraceCG))
+               {
+               printInst();
+               }
+            break;
+            }
+         }
+
+      moveInstr = true;
+      if(moveInstr)
+         _cursor = _cursor->getNext();
+      }
+
+   if (comp()->getOption(TR_TraceCG))
+      printInfo("\n\n");
+   }
+
+bool
+TR_S390PreRAPeephole::LoadStoreReduction(TR::InstOpCode::Mnemonic storeOpCode, uint16_t size)
+   {
+   if (_cursor->getNext()->getOpCodeValue() == storeOpCode)
+      {
+      TR::S390RXInstruction* loadInst = static_cast<TR::S390RXInstruction*> (_cursor);
+      TR::S390RXInstruction* storeInst = static_cast<TR::S390RXInstruction*> (_cursor->getNext());
+
+      if ( loadInst->getMemoryReference()->getSymbolReference() && loadInst->getMemoryReference()->getSymbolReference()->isUnresolved() ||
+          storeInst->getMemoryReference()->getSymbolReference() && storeInst->getMemoryReference()->getSymbolReference()->isUnresolved())
+         {
+         return false;
+         }
```

Comment:
I think this deserves a one liner comment why it is necessary as it is non-obvious at first glance.

Reply:
79eadfb